web3/middleware/filter.py

```diff
@@ -0,0 +1,326 @@
+      StopIteration
+    >>>
+    """
+    _last = None
+
+    is_bounded_range = (
+        to_block is not None and
+        to_block is not 'latest'
+    )
+
+    while True:
+        latest_block = w3.eth.blockNumber
+        if is_bounded_range and latest_block > to_block:
+            return
+        #  No new blocks since last iteration.
+        if _last is not None and _last == latest_block:
+            yield None
+        else:
+            yield latest_block
+        _last = latest_block
+
+
+def iter_latest_block_ranges(w3, from_block, to_block=None):
+    """Returns an iterator unloading ranges of available blocks
+
+    starting from `fromBlock` to the latest mined block,
+    until reaching toBlock. e.g.:
+
+
+    >>> blocks_to_filter = iter_latest_block_ranges(w3, 0, 50)
+    >>> next(blocks_to_filter)  # latest block number = 11
+    (0, 11)
+    >>> next(blocks_to_filter)  # latest block number = 45
+    (12, 45)
+    >>> next(blocks_to_filter)  # latest block number = 50
+    (46, 50)
+    """
+    for latest_block in iter_latest_block(w3, to_block):
+        if latest_block is None:
+            yield (None, None)
+        elif from_block > latest_block:
+            yield (None, None)
+        else:
+            yield (from_block, latest_block)
+            from_block = latest_block + 1
+
+
+def drop_items_with_none_value(params):
+    return valfilter(lambda x: x is not None, params)
+
+
+def get_logs_multipart(
+        w3,
+        startBlock,
+        stopBlock,
+        address,
+        topics,
+        max_blocks):
+    """Used to break up requests to ``eth_getLogs``
+
+    The getLog request is partitioned into multiple calls of the max number of blocks
+    ``max_blocks``.
+    """
+    _block_ranges = block_ranges(startBlock, stopBlock, max_blocks)
+    for from_block, to_block in _block_ranges:
+        params = {
+            'fromBlock': from_block,
+            'toBlock': to_block,
+            'address': address,
+            'topics': topics
+        }
+        yield w3.eth.getLogs(
+            drop_items_with_none_value(params))
+
+
+class RequestLogs:
+    def __init__(
+            self,
+            w3,
+            from_block=None,
+            to_block=None,
+            address=None,
+            topics=None):
+
+        self.address = address
+        self.topics = topics
+        self.w3 = w3
+        if from_block is None or from_block == 'latest':
+            self._from_block = w3.eth.blockNumber + 1
+        else:
+            self._from_block = from_block
+        self._to_block = to_block
+        self.filter_changes = self._get_filter_changes()
+
+    @property
+    def from_block(self):
+        return self._from_block
+
+    @property
+    def to_block(self):
+        if self._to_block is None:
+            to_block = self.w3.eth.blockNumber
+
+        elif self._to_block == 'latest':
+            to_block = self.w3.eth.blockNumber
+
+        else:
```

Comment:
*nit* I'm not generally a fan of this gap before the else or elif statement. Up to you, though.